internal/safebuffer/safereader/safereader.go

```diff
@@ -0,0 +1,160 @@
+package safereader
+
+import (
+	"bytes"
+	"io"
+	"log/slog"
+	"sync"
+)
+
+// Reader is a goroutine safe io.ReadCloser that makes initial data available to
+// callers as well as new data that arrives later via channel.
+type Reader struct {
+	mu        sync.Mutex
+	buf       *bytes.Buffer
+	close     func()
+	closed    <-chan struct{}
+	dataCh    <-chan []byte
+	dataAvail chan struct{}
+	closeOnce sync.Once
```

Comment:
There are a lot of concurrency control structures here too, I wonder if we could simplify.